### PR TITLE
Relationships api

### PR DESCRIPTION
**Pull request by sanderland**

### CODE REVIEW

File: cognite/client/_api/relationships.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+            Get relationship by external id::
+
+                >>> from cognite.client import CogniteClient
+                >>> c = CogniteClient()
+                >>> res = c.relationships.retrieve(external_id="1")
+        """
+        return self._retrieve_multiple(external_ids=external_id, wrap_ids=True)
+
+    def retrieve_multiple(self, external_ids: Optional[List[str]] = None) -> RelationshipList:
```
Comment (erlendvollset) [inline cognite/client/_api/relationships.py]: `external_ids` shouldn't be optional.

File: cognite/client/data_classes/relationships.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        raise ValueError("Invalid source or target '{}' of type {} in relationship".format(target, target.__class__))
+
+
+class RelationshipUpdate(CogniteUpdate):
+    pass
+
+
+class RelationshipList(CogniteResourceList):
+    _RESOURCE = Relationship
+    _UPDATE = RelationshipUpdate
```
Comment (erlendvollset) [inline cognite/client/data_classes/relationships.py]: You can remove the _UPDATE attribute and RelationshipUpdate class. Just need to set `_ASSERT_CLASSES = False`

File: cognite/client/data_classes/__init__.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 )
 from cognite.client.data_classes.login import LoginStatus
 from cognite.client.data_classes.raw import Database, DatabaseList, Row, RowList, Table, TableList
+from cognite.client.data_classes.relationships import Relationship, RelationshipList, RelationshipUpdate
```
Comment (erlendvollset) [inline cognite/client/data_classes/__init__.py]: Remove RelationshipUpdate here aswell.

File: cognite/client/_api/relationships.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+                >>> from cognite.client import CogniteClient
+                >>> from cognite.client.data_classes import Relationship
+                >>> c = CogniteClient()
+                >>> assets = c.assets.retrieve_multiple(id=[1,2,3])
+                >>> flowrel1 = Relationship(external_id="flow_1",source=assets[0],target=assets[1] ,relationship_type="flowsTo",confidence=0.1,data_set="ds_flow")
+                >>> flowrel2 = Relationship(external_id="flow_2",source=assets[1],target=assets[2] ,relationship_type="flowsTo",confidence=0.1,data_set="ds_flow")
+                >>> res = c.relationships.create([flowrel1,flowrel2])
+        """
+        utils._auxiliary.assert_type(relationship, "relationship", [Relationship, list])
```
Comment (erlendvollset) [inline cognite/client/_api/relationships.py]: Should we let this be a `RelationshipList` aswell?

File: cognite/client/_api/relationships.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
+from typing import *
+
+from cognite.client import utils
+from cognite.client._api_client import APIClient
+from cognite.client.data_classes import Relationship, RelationshipList
+from cognite.client.exceptions import CogniteAPIError
```
Comment (erlendvollset) [inline cognite/client/_api/relationships.py]: Unused?

File: cognite/client/_api/relationships.py (hunk 1 above)
Comment (sanderland) [reply cognite/client/_api/relationships.py]: meet my friends copy and paste ;) fixed

File: cognite/client/data_classes/relationships.py (hunk 2 above)
Comment (sanderland) [reply cognite/client/data_classes/relationships.py]: I prefer this as a todo marker

File: cognite/client/data_classes/__init__.py (hunk 3 above)
Comment (sanderland) [reply cognite/client/data_classes/__init__.py]: sure

File: cognite/client/_api/relationships.py (hunk 4 above)
Comment (sanderland) [reply cognite/client/_api/relationships.py]: I thought we generally didn't have a solution for this (as in - cleaning retrieved objects and copying them)

File: cognite/client/_api/relationships.py (hunk 5 above)
Comment (sanderland) [reply cognite/client/_api/relationships.py]: yep

File: cognite/client/_api/relationships.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            external_id (str): External ID
+
+        Returns:
+            Optional[Relationship]: Requested relationship or None if it does not exist.
+
+        Examples:
+
+            Get relationship by external id::
+
+                >>> from cognite.client import CogniteClient
```
Comment (erlendvollset) [inline cognite/client/_api/relationships.py]: Examples should import CogniteClient from experimental namespace, and be added to test_docstring_examples.py

File: tests/tests_unit/test_api/test_relationships.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            assert mock_rel_response.calls[0].response.json()["items"] == rel.dump(camel_case=True)
+
+    def test_delete_single(self, mock_rel_response):
+        res = REL_API.delete(external_id="a")
+        assert {"items": [{"externalId": "a"}]} == jsgz_load(mock_rel_response.calls[0].request.body)
+        assert res is None
+
+    def test_delete_multiple(self, mock_rel_response):
+        res = REL_API.delete(external_id=["a"])
+        assert {"items": [{"externalId": "a"}]} == jsgz_load(mock_rel_response.calls[0].request.body)
```
Comment (erlendvollset) [inline tests/tests_unit/test_api/test_relationships.py]: Should make this check for all the tests. Only testing that the response is correctly parsed, and not that the correct request is sent for the other tests.

**Pull request merged**